### PR TITLE
JBPM-5953: Add support for Service Task custom property values to Dat…

### PR DESCRIPTION
**Pull request by jlindop**
…a I/O editor

### CODE REVIEW

**Comment by tsurdilo** [comment]
@jlindop Hi, I ran your branch locally but seems nothing has changed (at least to me) from what was there before. Here is the update to the Email workitem i put in wid:

"parameterValues" : [
       "From" : "A,B,C",   
       "To" : "X,Y,Z",
       "Subject" : "one,two,three,four,five,six,seven,eight,nine,ten,eleven,twelve"
    ],

when i open the assignments editor and edit for example the "From" data input i don't see A B C options in the list. Do you think I am doing something wrong here?

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/bpmn2/impl/Bpmn2JsonUnmarshaller.java [hunk 1]
```diff
@@ -5990,15 +5990,8 @@ protected void applyTaskProperties(Task task,
...
-                        // check if custom workitem property has been updated
-                        if (isCustomElement(properties.get("tasktype"),
-                                            preProcessingData) &&
-                                (properties.get(fromPart.toLowerCase()) != null && properties.get(fromPart.toLowerCase()).length() > 0)) {
-                            fromExpression.setBody(properties.get(fromPart.toLowerCase()));
-                        } else {
-                            String replacer = decodeAssociationValue(assignmentParts[1]);
-                            fromExpression.setBody(wrapInCDATABlock(replacer));
-                        }
+                        String replacer = decodeAssociationValue(assignmentParts[1]);
```
Comment (tsurdilo) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/bpmn2/impl/Bpmn2JsonUnmarshaller.java]: This breaks the update of custom elements when "customEditor" : "true" is used in the wid. unless we completely disable this feature i think we should not break it :)

File: jbpm-designer-client/src/main/resources/org/jbpm/designer/public/stencilsets/bpmn2.0jbpm/stencildata/bpmn2.0jbpm.orig [hunk 2]
```diff
@@ -3812,8 +3812,8 @@
 					"optional":true
 					 $if(workitemDefs.(k).customEditor)$
 					,
-					"hidden":true,
-					"visible":false
+					"hidden":false,
+					"visible":true
```
Comment (tsurdilo) [inline jbpm-designer-client/src/main/resources/org/jbpm/designer/public/stencilsets/bpmn2.0jbpm/stencildata/bpmn2.0jbpm.orig]: why is this needed?

**Comment by tsurdilo** [comment]
@jlindop looks good (see comments for some questions/ issues).
I just either don't know how to test it or there is something going on because i do not see the default values when a comma-separated string is used for parameterValues in the wid. Let me know if you can what i might be doing wrong here. thanks.

File: jbpm-designer-backend/src/main/java/org/jbpm/designer/bpmn2/impl/Bpmn2JsonUnmarshaller.java (hunk 1 above)
Comment (jlindop) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/bpmn2/impl/Bpmn2JsonUnmarshaller.java]: The change I've added to propertywindow.js (and function definitions in dataioeditor.js) ensure that it isn't broken. When the user sets a custom property, ORYX.DataIOEditorUtils.setAssignmentsPropertyForCustomAssignment is called which updates the 'assignments' property with the new custom property value setting. When the BP is serialized, the values for custom properties are taken from the 'assignments' property rather than from the custom properties themselves.

Comment (tsurdilo) [inline jbpm-designer-backend/src/main/java/org/jbpm/designer/bpmn2/impl/Bpmn2JsonUnmarshaller.java]: @jlindop i tried used the per-property dropdowns when customeditor:true is set and it did not persist the values in my local testing but i can try again no problem.

File: jbpm-designer-client/src/main/resources/org/jbpm/designer/public/stencilsets/bpmn2.0jbpm/stencildata/bpmn2.0jbpm.orig (hunk 2 above)
Comment (jlindop) [inline jbpm-designer-client/src/main/resources/org/jbpm/designer/public/stencilsets/bpmn2.0jbpm/stencildata/bpmn2.0jbpm.orig]: This means that even when a service task has  customEditor set to true, the 'assignments' property is displayed, and the task has the button to bring up the DataIOEditor. Values for custom properties can be edited either in the custom property field or in the DataIOEditor, and the two are kept in sync. For example, if you update a custom property in the DataIOEditor, when the dialog is closed, the property field is updated with the new value.

**Comment by jlindop** [comment]
@tsurdilo   The way I've implemented this, you still need to have "customEditor" : "true" set to enable parameterValues. I tried your example above for the email task and after adding "customEditor" : "true" it works fine.

File: jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorViewImplTest.java [hunk 3]
```diff
@@ -169,6 +175,25 @@ public void testSetPossibleOutputAssignmentsVariables() {
...
+        customAssignmentsProperties.put("To",
+                                        Arrays.asList(new String[]{"Tom"}));
+        customAssignmentsProperties.put("CC",
+                                        Arrays.asList(new String[]{"James", "John"}));
+
+        view.setCustomAssignmentsProperties(customAssignmentsProperties);
+        verify(inputAssignmentsWidget).setCustomAssignmentsProperties(mapCaptor.capture());
+        assertEquals(3,
+                     mapCaptor.getValue().keySet().size());
+        assertEquals(customAssignmentsProperties.get("CC"),
```
Comment (jomarko) [inline jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorViewImplTest.java]: I think there could be assertion also for "From" and "To"

File: jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorTest.java [hunk 4]
```diff
@@ -257,6 +263,23 @@ public void testProcessVariables() {
...
+        customAssignmentsProperties.put("TruckType",
+                                        Arrays.asList(new String[]{"Hiace", "Mazda"}));
+        customAssignmentsProperties.put("To",
+                                        Arrays.asList(new String[]{"Donna", "Diana", "Julia"}));
+
+        ioEditor.setCustomAssignmentsProperties(customAssignmentsProperties);
+        verify(ioEditorView).setCustomAssignmentsProperties(mapCaptor.capture());
+        assertEquals(2,
+                     mapCaptor.getValue().keySet().size());
+        assertEquals(customAssignmentsProperties.get("TruckType"),
```
Comment (jomarko) [inline jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorTest.java]: I think there could be assertion also for "To"

File: jbpm-designer-client/src/main/java/org/jbpm/designer/client/shared/AssignmentData.java [hunk 5]
```diff
@@ -351,6 +356,68 @@ protected void setDisallowedPropertyNames(String disallowedPropertyNames) {
...
+                if (!propValues.isEmpty()) {
+                    Iterator<String> itPropValues = propValues.iterator();
+                    while (itPropValues.hasNext()) {
+                        String propValue = itPropValues.next();
+                        if (!propValue.isEmpty()) {
+                            sb.append(propValue).append(';');
+                        }
+                    }
+                }
+                sb.append(',');
```
Comment (jomarko) [inline jbpm-designer-client/src/main/java/org/jbpm/designer/client/shared/AssignmentData.java]: Just question here. This line appends the comma separator also after the last assignment. The method `setCustomAssignmentProperties` can handle this. Is this fine also for the javascript code?

File: jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorViewImplTest.java (hunk 3 above)
Comment (jlindop) [inline jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorViewImplTest.java]: Thanks - I'll add these.

File: jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorTest.java (hunk 4 above)
Comment (jlindop) [inline jbpm-designer-client/src/test/java/org/jbpm/designer/client/popup/ActivityDataIOEditorTest.java]: I'll add this. Thanks.

File: jbpm-designer-client/src/main/java/org/jbpm/designer/client/shared/AssignmentData.java (hunk 5 above)
Comment (jlindop) [inline jbpm-designer-client/src/main/java/org/jbpm/designer/client/shared/AssignmentData.java]: This implementation matches the javascript implementation here:
https://github.com/kiegroup/jbpm-designer/pull/611/files#diff-5b7904f8921bdfb9b4529d5bd6be3f36R218

The value of this string is only used by the Java code - it's passed from Javascript to Java when the DataIOEditor is invoked. On the return trip, it's only there for completeness of the serialization of the AssignmentData.

Comment (jomarko) [inline jbpm-designer-client/src/main/java/org/jbpm/designer/client/shared/AssignmentData.java]: thanks

**Comment by tsurdilo** [comment]
@jlindop i tested again after watching your video and i think its pretty close thanks! Take a look at https://streamable.com/nvgdr if you have the time- if you set the standalone props without touching the Data io editor the data inputs double up (from 4 to 8 in this case). I'll debug and see what can be done.

**Comment by jdlindop** [comment]
@tsurdilo  I've emailed you a fix for the bug in your video (:

**Comment by tsurdilo** [comment]
i needed to make some minor fixes / changes but cannot amend this pr. will close this pr and open a new one that includes all this code and my changes.
